In [1]:
#Importing necessary Libraries
import pandas as pd

In [2]:
#Reading Data
df = pd.read_csv('vehicles_us.csv')
#Looking at Data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


12 Columns, 51525 Rows.
Mandatory columns seem to be (Price, Model, Condition, Fuel(economy?), Transmission, Type, Date Posted, Days Listed) 8/12
Columns with missing data are (Model Year, Cylinders, Odometer, Paint Color, Is 4WD) 4/12

Model Year will be changed to integer.
Fuel being an object will be further investigated.
Is 4WD should be a boolean.
Price and Cylinders will be investigated for better data type.
Date Posted will be converted to Date data type.

In [3]:
#Getting a sample of data
df.sample(20)

price  model_year                model  condition  cylinders    fuel  \
38728  12995      2016.0    volkswagen passat       good        4.0     gas   
8946    3300      2002.0           ford f-150       good        8.0     gas   
45889   5900      2011.0        nissan maxima  excellent        NaN     gas   
49198   6450      2007.0         ford mustang  excellent        NaN     gas   
9080   14925      2015.0             ram 1500  excellent        8.0     gas   
30638   2999      2004.0    toyota highlander       good        6.0     gas   
41517  29990      2016.0   chevrolet suburban   like new        8.0     gas   
7693    6900      2013.0         nissan rogue       good        4.0     gas   
41367   2800      2006.0    volkswagen passat   like new        4.0     gas   
7727    4500      2001.0           ford f-250       fair        8.0  diesel   
24602   4995         NaN       toyota 4runner  excellent        6.0     gas   
40954  12677      2015.0        jeep cherokee  excellent        4.0     gas   
41399  15900      2001.0             ram 2500  excellent        6.0  diesel   
24462  17200      2015.0          honda pilot  excellent        NaN     gas   
34811  22997      2016.0           ford f-150       good        6.0     gas   
41686   6200      2009.0       ford fusion se       good        4.0     gas   
31879  17495      2005.0             ram 2500  excellent        6.0  diesel   
50236   5900      2007.0  chevrolet silverado       good        6.0     gas   
46531  24900      2013.0        ford f-350 sd  excellent        8.0  diesel   
44644    500      2000.0        nissan altima       fair        4.0     gas   

       odometer transmission    type paint_color  is_4wd date_posted  \
38728       NaN    automatic   sedan        grey     NaN  2018-07-15   
8946        NaN    automatic  pickup       white     1.0  2018-12-30   
45889  141000.0    automatic   sedan         red     NaN  2018-05-10   
49198       NaN       manual   coupe         NaN     NaN  2018-08-26   
9080    78779.0    automatic   truck       white     NaN  2018-11-22   
30638  214258.0    automatic     SUV       black     NaN  2018-12-22   
41517  122987.0    automatic     SUV      silver     1.0  2018-07-11   
7693        NaN    automatic     SUV       black     NaN  2018-06-10   
41367  141000.0    automatic   sedan        blue     NaN  2018-09-03   
7727   265000.0    automatic   truck       black     1.0  2018-11-06   
24602  262564.0    automatic     SUV       green     NaN  2018-12-20   
40954   47000.0    automatic     SUV         red     NaN  2018-06-14   
41399  250000.0    automatic  pickup         NaN     1.0  2019-01-04   
24462       NaN    automatic     SUV        grey     1.0  2018-09-30   
34811   26001.0    automatic   truck         NaN     NaN  2019-04-13   
41686   72000.0    automatic   sedan      silver     NaN  2018-08-06   
31879  179574.0    automatic   truck        blue     1.0  2019-03-28   
50236  158000.0    automatic   truck       white     NaN  2019-03-24   
46531  215359.0    automatic   truck       black     1.0  2019-03-12   
44644  270000.0    automatic   sedan      purple     NaN  2019-02-20   

       days_listed  
38728          110  
8946            64  
45889           70  
49198           34  
9080            25  
30638           36  
41517           27  
7693            69  
41367           31  
7727            63  
24602           23  
40954           36  
41399           14  
24462           23  
34811           19  
41686           18  
31879           53  
50236           51  
46531           13  
44644           25

In [4]:
#Printing missing values
print('Missing values:')
print(df.isna().sum())
print()

#Printing unique values in the 'is_4wd' column
print('Unique values in "is_4WD":')
print(df['is_4wd'].unique())

Missing values:
price               0
model_year       3619
model               0
condition           0
cylinders        5260
fuel                0
odometer         7892
transmission        0
type                0
paint_color      9267
is_4wd          25953
date_posted         0
days_listed         0
dtype: int64

Unique values in "is_4WD":
[ 1. nan]


### Fixing Data

In [5]:
#Changing 'date_posted' to datetime data type
df['date_posted'] = pd.to_datetime(df['date_posted'])

#Chagning 'model_year' to int data type
df['model_year'] = df['model_year'].fillna(-1).astype('int64')

#Filling missing values in 'is_4wd' and changing to int data type
df['is_4wd'] = df['is_4wd'].fillna(0).astype('int64')
df.sample(20)

price  model_year                          model  condition  cylinders  \
30839  18990        2011       chevrolet silverado 1500       good        8.0   
43568  17990        2015                     ford f-150       good        6.0   
1004    1000          -1                    ford taurus       good        6.0   
29575  16995        2014            jeep grand cherokee  excellent        6.0   
35470   5995        2010                   honda accord       good        6.0   
37034   7900        2006                   dodge dakota       good        8.0   
10149  24600        2014  chevrolet silverado 1500 crew       good        6.0   
28607   5995        2013       chevrolet silverado 1500  excellent        6.0   
11294  12950        2011                     ford f-150  excellent        8.0   
20654  15400        2017                    ford fusion  excellent        4.0   
30604  13800        2011                     ford f-150  excellent        NaN   
48835   6900        2011            dodge grand caravan       good        6.0   
14043  14588        2013                  ford explorer       good        6.0   
5045   13500        2017                toyota camry le  excellent        4.0   
35100   8477        2011                 subaru outback       good        4.0   
40669   7500        1998            chevrolet silverado       good        6.0   
43455  15997        2010               chevrolet camaro   like new        NaN   
38499   2900        2003                  honda odyssey       good        NaN   
36994   3495        2011                chevrolet cruze  excellent        4.0   
27810   5495        2006              volkswagen passat   like new        4.0   

      fuel  odometer transmission    type paint_color  is_4wd date_posted  \
30839  gas   47143.0    automatic   truck       white       1  2019-03-01   
43568  gas  129000.0    automatic   truck       white       1  2018-09-12   
1004   gas   81000.0    automatic   sedan        blue       0  2019-03-11   
29575  gas   84000.0    automatic     SUV      silver       1  2019-03-13   
35470  gas  122879.0    automatic   sedan        grey       0  2019-02-10   
37034  gas  141200.0    automatic   truck      silver       1  2018-06-20   
10149  gas   85048.0    automatic  pickup         NaN       1  2019-01-05   
28607  gas  184000.0    automatic   truck       white       0  2018-06-13   
11294  gas  149014.0    automatic   truck      silver       1  2019-02-19   
20654  gas   45048.0    automatic   sedan        grey       0  2018-08-23   
30604  gas  158700.0    automatic   truck      silver       1  2018-11-20   
48835  gas  143256.0    automatic     van         NaN       0  2018-06-24   
14043  gas  153445.0    automatic     SUV       white       1  2019-01-20   
5045   gas   48569.0    automatic   sedan      silver       1  2018-10-21   
35100  gas  139610.0    automatic   wagon        blue       1  2018-05-23   
40669  gas  115000.0       manual   truck         NaN       1  2018-07-16   
43455  gas       NaN        other   coupe      orange       0  2018-10-17   
38499  gas       NaN    automatic     van        blue       0  2018-12-11   
36994  gas  104000.0       manual   sedan        blue       0  2018-08-25   
27810  gas  129000.0    automatic   sedan       green       0  2018-12-07   

       days_listed  
30839           19  
43568           28  
1004            52  
29575           60  
35470            6  
37034           57  
10149           33  
28607            9  
11294          147  
20654            9  
30604           43  
48835           64  
14043           39  
5045             6  
35100           21  
40669          102  
43455           31  
38499            1  
36994           23  
27810           34

### Enrich Data


In [6]:
df['manufacturer'] = df['model'].apply(lambda x: x.split()[0])
df.sample(20)

price  model_year                     model  condition  cylinders  \
9547    7500        2007         toyota highlander        new        6.0   
27782   3995        2008              chrysler 300       good        6.0   
12085   6400        2010               honda civic  excellent        4.0   
48421   6995        2013         chevrolet equinox  excellent        4.0   
50606  26400        2016  ford mustang gt coupe 2d       good        8.0   
41993  26987        2014                 ford f150  excellent        6.0   
15160  12900          -1               ford fusion  excellent        4.0   
13591   1100        2008              toyota prius       fair        4.0   
24289   4250        2013             ford focus se   like new        4.0   
31890   4995        2007            hyundai sonata  excellent        4.0   
30310  21995        2017                 ford edge  excellent        6.0   
13855   6990          -1           chevrolet tahoe  excellent        8.0   
11941   4980        2012               ford fusion  excellent        4.0   
16222   8985        2013                honda cr-v  excellent        4.0   
46018   6995        2009          chevrolet malibu  excellent        4.0   
32364   8995        2015              chrysler 200  excellent        4.0   
13725   7900        2012              ford mustang       good        6.0   
48331  35000        2016                gmc sierra  excellent        8.0   
8041    4995        2002           chevrolet tahoe  excellent        8.0   
24643   3950        2008                ford f-150       good        8.0   

         fuel  odometer transmission       type paint_color  is_4wd  \
9547      gas  172000.0    automatic        SUV      silver       0   
27782     gas  184742.0    automatic      sedan        blue       0   
12085     gas  112000.0    automatic      coupe        grey       0   
48421     gas  133142.0    automatic        SUV      silver       0   
50606     gas   16890.0        other      coupe       white       0   
41993     gas   96070.0    automatic     pickup        blue       1   
15160     gas       NaN    automatic      sedan       white       0   
13591  hybrid  156000.0    automatic  hatchback        blue       0   
24289     gas  113500.0    automatic      sedan         NaN       0   
31890     gas   61996.0    automatic      sedan        grey       0   
30310     gas   17876.0    automatic        SUV        grey       1   
13855     gas  147485.0    automatic        SUV      silver       1   
11941     gas   79875.0    automatic      sedan       white       0   
16222     gas   99928.0    automatic        SUV       black       1   
46018     gas   94200.0    automatic      sedan      custom       0   
32364     gas   90789.0    automatic      sedan        blue       0   
13725     gas   77000.0       manual      coupe       black       0   
48331     gas   39500.0    automatic      truck       black       1   
8041      gas       NaN    automatic        SUV       green       1   
24643     gas       NaN    automatic      truck       white       0   

      date_posted  days_listed manufacturer  
9547   2018-11-16           30       toyota  
27782  2019-03-15           62     chrysler  
12085  2018-06-14           68        honda  
48421  2018-07-07           18    chevrolet  
50606  2019-02-01           79         ford  
41993  2018-08-04           65         ford  
15160  2019-01-09           29         ford  
13591  2018-11-27           41       toyota  
24289  2019-04-08           26         ford  
31890  2018-08-19           11      hyundai  
30310  2018-09-11           23         ford  
13855  2018-05-02           22    chevrolet  
11941  2018-06-22           26         ford  
16222  2018-05-13           33        honda  
46018  2018-10-22           42    chevrolet  
32364  2018-07-21          105     chrysler  
13725  2019-01-03          102         ford  
48331  2018-08-05          100          gmc  
8041   2019-02-09           53    chevrolet 

In [27]:
manu_year = df.groupby(['model_year', 'manufacturer', 'condition'])['price'].mean().reset_index()
manu_year[manu_year['model_year'] == 1999]

model_year manufacturer  condition         price
481        1999        acura       fair   1200.000000
482        1999        acura       good   2947.500000
483        1999     cadillac       fair    500.000000
484        1999    chevrolet  excellent   7811.529412
485        1999    chevrolet       fair   2502.352941
486        1999    chevrolet       good   5534.928571
487        1999    chevrolet   like new  13916.666667
488        1999    chevrolet    salvage    800.000000
489        1999        dodge  excellent   1550.000000
490        1999        dodge       fair   1437.500000
491        1999        dodge       good   2975.000000
492        1999         ford  excellent   8827.568966
493        1999         ford       fair   3252.173913
494        1999         ford       good   5737.523810
495        1999         ford   like new   9829.000000
496        1999          gmc  excellent   4897.500000
497        1999          gmc       fair   4750.000000
498        1999          gmc       good   3814.300000
499        1999          gmc    salvage    800.000000
500        1999        honda  excellent   2590.000000
501        1999        honda       fair   1035.857143
502        1999        honda       good   2269.900000
503        1999         jeep  excellent   8288.000000
504        1999         jeep       fair   1643.125000
505        1999         jeep       good   5491.800000
506        1999         jeep   like new   5545.000000
507        1999         jeep    salvage   1000.000000
508        1999       nissan  excellent   2225.000000
509        1999       nissan       good  39599.500000
510        1999          ram  excellent  11316.642857
511        1999          ram       fair   4342.142857
512        1999          ram       good   7415.280000
513        1999          ram        new  15300.000000
514        1999       subaru  excellent   3350.000000
515        1999       subaru       fair   2450.000000
516        1999       subaru       good   3235.666667
517        1999       toyota  excellent   5359.684211
518        1999       toyota       fair   1717.727273
519        1999       toyota       good   3695.690476
520        1999       toyota   like new   4698.250000
521        1999       toyota    salvage   1666.666667
522        1999   volkswagen       fair   1500.000000
523        1999   volkswagen       good   2031.666667